pepper/apps/prover_knows_hash_preimage.c

```diff
@@ -0,0 +1,163 @@
+#include <prover_knows_hash_preimage.h>
+
+/**
+ * Algorithm Description:
+ *
+ * 1) Verifier sets up the P and V keys:
+ *      ./pepper_compile_and_setup_V.sh prover_knows_hash_preimage prover_knows_hash_preimage.vkey prover_knows_hash_preimage.pkey
+ *
+ * 2) Prover claims that she knows the preimage to that hash        (located in the exo0 file)
+ *      ./pepper_compile_and_setup_P.sh prover_knows_hash_preimage
+ *
+ * 3) Verifier generates and provides the input_hash (which is a sha256 hash)     (located in the prover_knows_hash_preimage.inputs file)
+ *      bin/pepper_verifier_prover_knows_hash_preimage gen_input prover_knows_hash_preimage.inputs
+ *
+ * 4) Prover generates a proof (prover_knows_hash_preimage.proof) that he knows the preimage
+ *      bin/pepper_prover_prover_knows_hash_preimage prove prover_knows_hash_preimage.pkey prover_knows_hash_preimage.inputs prover_knows_hash_preimage.outputs prover_knows_hash_preimage.proof
+ *
+ * 5) Finally, Verifier checks the computation without ever knowing the preimage
+ *      bin/pepper_verifier_prover_knows_hash_preimage verify prover_knows_hash_preimage.vkey prover_knows_hash_preimage.inputs prover_knows_hash_preimage.outputs prover_knows_hash_preimage.proof
+**/
+
+
+void compute(struct In *input, struct Out *output) {
+    uint32_t i;
+    preimage_t preimage[1];                         /* prover will fill it with his private preimage */
+    uint8_t prover_hash[SHA256_BLOCK_SIZE];         /* this is the hash of private preimage */
+
+    output->prover_knows_preimage = 1;
```

Comment:
I think this variable should be initialized to 0 and the loop below should check if all bytes of hash are equal to prover's hash. I know it is the same but I think is more clear. We assume that the prover does not know the preimage and we want a proof that it knows it.